Implement light and dark themes for design picker step

alshakero opened the pull request:
#### Changes proposed in this Pull Request

* This PR adds binary (light and dark) theming ability to the design picker. 
* It adds `theme` prop to DesignPicker.
* It adjusts the text color and the border colors according to this prop's value.

### Testing instructions

#### Testing dark mode

- Visit [/start/with-design-picker](https://calypso.live/start/with-design-picker?branch=update/design-picker-step-styling) and reach design picker step.
	- [ ] Design names should be in white.
	- [ ] Design preview images border should be gray (`--studio-gray-40`) when idle, white when active or hovered.

#### Testing light mode

- Visit [/new](https://calypso.live/new?branch=update/design-picker-step-styling) and reach design picker step.
	- [ ] Design names should be in gray (`--studio-gray-40`).
	- [ ] Design preview images border should be gray (`--studio-gray-5`) when idle, blue when active or hovered (`--studio-blue-20`).

Fixes: https://github.com/Automattic/wp-calypso/issues/51401

alshakero committed:
Adopt exPlat changes in with-design-picker flow

alshakero committed:
Implement light and dark theme for design picker step

ciampo commented on client/signup/controller.js:
```diff
@@ -99,10 +99,11 @@ export const removeP2SignupClassName = function () {
 	redirectTests( context, next ) {
 		const currentFlowName = getFlowName( context.params );
-		currentFlowName === 'onboarding' && loadExperimentAssignment( 'refined_reskin_v1' );
+		[ 'onboarding', 'with-design-picker' ].includes( currentFlowName ) &&
```
Are we enabling this experiment for the `with-design-picker` flow only for us to test it in this PR, or will this change be permanent ?

razvanpapadopol replied:
It shouldn't be permanent since we are not planning to overlap the two experiments. See pbxNRc-LC-p2#comment-1982

But I think it's a good way to make sure the design picker is compatible with a light theme.

alshakero replied:
Addressed in a call

ciampo commented on client/signup/steps/user/style.scss:
```diff
@@ -1,4 +1,5 @@
-body.is-section-signup.is-white-signup .signup.is-onboarding {
+body.is-section-signup.is-white-signup .signup.is-onboarding,
+body.is-section-signup.is-white-signup .signup.is-with-design-picker {
```
Why do we need to apply these changes to the `user` step?

alshakero replied:
Addressed in a call

ciampo commented on packages/design-picker/src/components/index.tsx:
```diff
@@ -40,11 +41,12 @@ const DesignPicker: React.FC< Props > = ( {
 
 	return (
-		<div className="design-picker">
+		<div className={ classnames( 'design-picker', `theme-${ theme }` ) }>
```
Don't have a strong opinion on className conventions, but it seems like we're using some sort of BEM for design picker

What about 
```suggestion
		<div className={ classnames( 'design-picker', `design-picker--theme-${ theme }` ) }>
```

alshakero committed:
Remove with-design-picker selector from user signup

alshakero commented:
Hi @ollierozdarz! Can you give the design picker here a quick design review please?

### To see the dark mode: 

1. Go to https://hash-e838092ef50489a33b6c1f999b224e792c8d38d8.calypso.live/start/with-design-picker and pick a domain, you should reach the design picker step then.
2. Please check the hovering state too.

### To see the light mode:
1. Go to the link stored here 29134-pb.
2. Assign yourself to "treatment".
3. Repeat the steps above from the dark mode.

ciampo commented on packages/design-picker/src/components/style.scss:
```diff
@@ -173,3 +173,10 @@
 		}
 	}
 }
+
+/// dark theme styles
+.design-picker.theme-dark {
+	.design-picker__option-name {
+		color: var( --studio-white );
+	}
+}
```
At the moment, the text color used for the "light" theme is the [default text color](https://github.com/Automattic/wp-calypso/pull/52162/files#diff-3c52a0567c030fc5e6767726be99f162eab5d06c2ee122cd1f8713b586c66733R160).

I wonder if it would be better to make that rule explicit for the light theme too. In my opinion, it would make the code more readable, and it would make sure that the text color CSS rules have the same specificity for both dark and light theme

```suggestion

/// dark theme styles
.design-picker.theme-dark {
	.design-picker__option-name {
		color: var( --studio-white );
	}
}

// light theme styles
.design-picker.theme-light {
	.design-picker__option-name {
		color: var( --studio-gray-40 );
	}
}
```

ciampo commented on packages/design-picker/src/components/style.scss:
```diff
@@ -173,3 +173,10 @@
 	}
 }
+
+/// dark theme styles
```
Nit: Extra `/`

```suggestion
// dark theme styles
```

ciampo reviewed (changes requested):
While waiting for @ollierozdarz  feedback, I had a first look at this PR

### Hover/focus styles

At the moment, the hover/focus `border-color` [is determined by a CSS variable called`--highlightColor`](https://github.com/Automattic/wp-calypso/blob/ac2d251ff54cf2e7807e537c4d8068e427ddac0c/packages/design-picker/src/components/style.scss#L32-L37). This variable is [defined for the `gutenboarding` flow](https://github.com/Automattic/wp-calypso/blob/f64e74fbca5741ec2daf0943823480d04fd7d092/client/landing/gutenboarding/onboarding-block/colors.scss), and it changes depending on external factors (like the vertical). This means that this variable is not defined in the context of the `/start` flow, which is why  **the thumbnail's `border-color` is black in the `/start/with-design-picker` flow.**

At this point, I believe that we need to agree on what should our approach be:

- should the `border-color` change with the dark/light theme? (at the moment, it isn't) If yes, to which colours?
- should the `border-color` be defined from outside the design picker package? (at the moment, it's defined from an external CSS variable)

A potential, non-disruptive solution that comes to mind would be:
- Move the `border-color` CSS rules to be specific for `dark` and `light` theme (like we would be doing for the text color)
- Stop using the`--highlightColor` CSS variable
- For the light theme, use the `--studio-blue-20` border-color
- For the dark theme, maybe use the `--studio-white` color? (would be good to hear @ollierozdarz 's feedback on this one)

### Missing integration tests

At the moment, this PR is still missing integration tests for the new `theme` prop (see [the `/packages/design-picker/src/__tests__`](https://github.com/Automattic/wp-calypso/tree/trunk/packages/design-picker/src/__tests__) folder)

### Integration with reskin experiment

Finally, regarding the code changes that enable the reskin experiment for the `/start/with-design-picker` flow:

- It's good to have these changes while reviewing the PR, as they give us a way to test both light and dark theme in the context of `/start` signup flow
- As discussed last week, though, these changes are probably not in the scope for this PR, and we may want to remove them before merging

alshakero committed:
Add theming int-test

ollierozdarz commented:
> For the light theme, use the --studio-blue-20 border-color
For the dark theme, maybe use the --studio-white color? (would be good to hear @ollierozdarz 's feedback on this one)

These sound great to me @ciampo. The current dark hover state on the _control_ background feels a little strange since it visually gives the impression that the design is receding into the background. The solid White would fix that nicely.

ollierozdarz reviewed (commented):
LGTM 👍

alshakero committed:
Remove exPlat modifications and address feedback

alshakero committed:
Tidy up integration test

alshakero commented:
@ciampo I addressed all feedback and rewrote the PR's description.

@ollierozdarz Hi again! I improvised a bit regarding the idle state of the border (without hovering) in dark mode. I used `--studio-gray-40`. You can take a look [here](https://calypso.live/start/with-design-picker?branch=update/design-picker-step-styling). Let me know what you think 🙏 .

ciampo commented on packages/design-picker/src/components/style.scss:
```diff
@@ -173,3 +165,36 @@
+	.design-picker__option-name {
+		color: var( --studio-gray-40 );
+	}
+	.design-picker__design-option {
```
Semi-nit, but:

similarly to how I [suggested moving the CSS rules the text color](https://github.com/Automattic/wp-calypso/pull/52162#discussion_r620098964) for both dark and light themes, I think that it would be better to have the `border-color` CSS rule for the light theme here too:

```suggestion
	.design-picker__design-option {
		.design-picker__image-frame {
			border-color: var( --studio-gray-5 );
		}
```

With this change applied, the CSS rule (currently at line 102) that says `border: 1px solid var( --studio-gray-5 );` could be replaced by `border: 1px solid;`

ciampo commented on packages/design-picker/src/__tests__/integration.test.tsx:
```diff
@@ -45,4 +46,18 @@ describe( '<DesignPicker /> integration', () => {
 		);
 	} );
+
+	( [ 'light', 'dark' ] as Props[ 'theme' ][] ).forEach( ( theme ) =>
```
Should we open a new issue to track improving these integration tests (ie. trying to check computed styles for borders and text colors)?

cc @StefanNieuwenhuis and @razvanpapadopol

razvanpapadopol replied:
I've opened https://github.com/Automattic/wp-calypso/issues/52370 to track this.

- created a new Milestone to keep track of DesignPicker stability issues like this:
https://github.com/Automattic/wp-calypso/milestone/354
- assigned also https://github.com/Automattic/wp-calypso/issues/51743 to the new milestone.

cc: @Automattic/luna

ciampo replied:
Awesome, thank you!

ciampo reviewed (approved):
I left a couple of comments, but after addressing them, this PR LGTM from the technical point of view.

We should wait for the final blessing from @ollierozdarz regarding the border color, as per @alshakero's previous comment.

ollierozdarz commented:
Looks great @alshakero @ciampo 🙂

alshakero committed:
Move theming styles together

razvanpapadopol commented on client/signup/steps/design-picker/index.jsx:
```diff
@@ -54,7 +54,14 @@ class DesignPickerStep extends Component {
+		return (
+			<DesignPicker
+				theme="dark"
+				isGridMinimal
```
Isn't this flag supposed to be used only for AnchorFM (and maybe FSE) flows?

In the default use case where we have many designs listed, it's probably better to show 3 rows instead of 2 on a larger screen:

- Production on a 1920px screen:

<img width="960" alt="Screenshot 2021-04-29 at 14 57 05" src="https://user-images.githubusercontent.com/14192054/116547478-98474f80-a8fb-11eb-8af0-970f085f6261.png">

- This branch on a 1920px screen:

<img width="958" alt="Screenshot 2021-04-29 at 14 56 40" src="https://user-images.githubusercontent.com/14192054/116547454-941b3200-a8fb-11eb-8ab4-f4656c3a5165.png">

cc: @ciampo

alshakero replied:
Yes it is. This PR was before yours that fixed the width. This was a workaround but now it's no longer needed!

alshakero committed:
Remove minimal flag
